core/solver/batch_cg_kernels.hpp

```diff
@@ -0,0 +1,161 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2021, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_CORE_SOLVER_BATCH_CG_KERNELS_HPP_
+#define GKO_CORE_SOLVER_BATCH_CG_KERNELS_HPP_
+
+
+#include <ginkgo/core/matrix/batch_csr.hpp>
+#include <ginkgo/core/matrix/batch_dense.hpp>
+
+#include <ginkgo/core/preconditioner/batch_preconditioner_types.hpp>
+#include <ginkgo/core/stop/batch_stop_enum.hpp>
+#include "core/log/batch_logging.hpp"
+
+namespace gko {
+namespace kernels {
+namespace batch_cg {
+
+
+/**
+ * Options controlling the batch Cg solver.
+ */
+template <typename RealType>
+struct BatchCgOptions {
+    preconditioner::batch::type preconditioner;
+    int max_its;
+    RealType rel_residual_tol;
+    RealType abs_residual_tol;
+    ::gko::stop::batch::ToleranceType tol_type;
+};
```

Comment:
Maybe we can unify this for all the batch solvers ?

Reply:
Actually, I'm not completely sure of that. We'd have unused parameters lying around for many solvers. Especially IDR has parameters that no one else needs. Maybe it's fine to have separate structs.

Reply:
Could we do something like- have a BatchSolverOptions struct which houses all common parameters, and then derive  Gmres and IDR Options struct from it and just add extra members while doing so?